keps/sig-storage/20191117-immutable-ephemeral-volumes.md

```diff
@@ -0,0 +1,196 @@
+
+**Note:** Any PRs to move a KEP to `implementable` or significant changes once it is marked `implementable` should be approved by each of the KEP approvers. If any of those approvers is no longer appropriate than changes to that list should be approved by the remaining approvers and/or the owning SIG (or SIG-arch for cross cutting KEPs).
+
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://github.com/kubernetes/enhancements/issues
+[kubernetes/kubernetes]: https://github.com/kubernetes/kubernetes
+[kubernetes/website]: https://github.com/kubernetes/website
+
+## Summary
+
+The most popular and the most convinient way of consuming Secrets and
+ConfigMaps by Pods is consuming it as a file. However, any update to
+a Secret or ConfigMap object is quickly (roughly within a minute)
+reflected in updates of the file mounted for *all* Pods consuming them.
+That means, that a bad update (push) of Secret and/or ConfigMap can
+very quickly break the whole appplication.
+
+The recommended approach for upgrading applications are obviously
+rolling upgrades. For Secrets/ConfigMaps this is translating to
+creating a new object and updating PodTemplate with the reference
+to it. However, that doesn't protect users from outages caused by
+accidental bad updates of existing Secrets and/or ConfigMaps.
+
+Moreover, the feature of updating Secrets/ConfigMaps for running
+Pods is also expensive from scalability/performance point of view.
+Every Kubelet has to watch (default) or periodically poll every
+unique Secret and ConfigMap referenced by any of the Pods it is
+running.
+
+## Motivation
+
+In this KEP, we are proposing to introduce an ability to specify
+that contents of a particular Secret/ConfigMap should be immutable
+for its whole lifetime. For those Secrets/ConfigMap, Kubelets will
+not be trying to watch/poll for changes to updated mounts for their
+Pods.
+Given there are a lot of users not really taking advantage of automatic
+updates of Secrets/ConfigMaps due to consequences described above, this
+will allow them to:
+- protect themselves better for accidental bad updates that could cause
+  outages of their applications
+- achieve better performance of their cluster thanks to significant
+  reduction of load on apiserver
+
+### Goals
+
+- introduce protection mechanism to avoid outages due to accidental
+  updates of existing Secrets/ConfigMaps
+- improve cluster performance by reducing load on Kubernetes control
+  plance (mostly kube-apiserver) consumed by not-utilized feature
+
+### Non-Goals
+
+- change the default behavior of consumption of Secrets/ConfigMaps
+
+## Proposal
+
+We propose to extend `ConfigMap` and `Secret` types with an additional
+field:
+```go
+  Immutable bool
```

Comment:
Could be:

```go
    Immutable *bool
```

Then we can develop a webhook admission to set users' Configmap/Secrets to be immutable as default

Reply:
`*bool` can differ user specified value or the golang default bool value.  And in Kubernetes, we can treat `nil` as false or set it to be `false` as default.

In our environment, if user does not set, our admission webhook will set Immutable to be `true` as default.

Reply:
Yes - it should be pointer changed.

Regarding webhook - yes we may do that, though it's more an extension, so I wouldn't put it into this proposal explicitly.